File: 994. Rotting Oranges.java
class Solution {// Time and space of O(nm)
    public int orangesRotting(int[][] grid) {
        //base case
        if(grid == null)
            return 0;
        int nrow= grid.length ;
        int ncol = grid[0].length ;
        int size = 0;
        // direction array for trackig index
        int[][] dirs ={{1,0},{0,1},{-1,0},{0,-1}};
        int freshOranges = 0;
        
        Queue<Pair<Integer,Integer>> queue = new LinkedList<>();
        // find all rotton oranges and add that to the queue
        for(int i = 0 ; i<nrow ; i++){
            for(int j =0 ; j< ncol ; j++){
                if(grid[i][j] == 2)
                    queue.add(new Pair(i,j));
                else if(grid[i][j] == 1)
                    freshOranges += 1;     
            }
        }
        if(freshOranges == 0) return 0;
        int time = -1;
        while(!queue.isEmpty()){
            size = queue.size();
            for(int i = 0; i<size ; i++){
                Pair<Integer,Integer> cell = queue.poll();
                int row = cell.getKey();
                int col = cell.getValue();
              for(int[] dir:dirs){
                int x = row + dir[0];
                int y = col + dir[1];
                if(isValid(x,y,grid)){
                    grid[x][y] = 2;
                    freshOranges-- ;
                    queue.add(new Pair(x,y));
                } 
               }
            }
            time +=1;
        }
        return freshOranges ==0? time : -1 ;
    }
    private boolean isValid(int x, int y, int[][] grid){
       int nrow= grid.length ;
       int ncol = grid[0].length ; 
       if(x>=0 && x<nrow && y>=0 && y<ncol && grid[x][y] == 1)
           return true;
       return false;
    }
}

File: 994. Rotting Oranges.py
"""
In a given grid, each cell can have one of three values:

the value 0 representing an empty cell;
the value 1 representing a fresh orange;
the value 2 representing a rotten orange.
Every minute, any fresh orange that is adjacent (4-directionally) to a rotten orange becomes rotten.

Return the minimum number of minutes that must elapse until no cell has a fresh orange.  If this is impossible, return -1 instead.

 

Example 1:



Input: [[2,1,1],[1,1,0],[0,1,1]]
Output: 4
Example 2:

Input: [[2,1,1],[0,1,1],[1,0,1]]
Output: -1
Explanation:  The orange in the bottom left corner (row 2, column 0) is never rotten, because rotting only happens 4-directionally.
Example 3:

Input: [[0,2]]
Output: 0
Explanation:  Since there are already no fresh oranges at minute 0, the answer is just 0.
"""

class Solution:
"""
    Run TIme Complexity: O(N*M)
    Space Complexity: O(N)
    
    
    Traverse through entire grid and count number of fresh fruits.
    Add all rotten fruits to queue.
    Now perform BFS and spoil all fresh fruits that are in surrounding of the rotten fruits.
"""
    def orangesRotting(self, grid: List[List[int]]) -> int:
        minutes = 0
        count = 0
        queue = deque()
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if(grid[i][j]) == 1: count += 1
                if(grid[i][j] == 2):
                    queue.append((i, j))
                        
        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]

        while queue:
            for i in range(len(queue)):
                i , j = queue.popleft()
                for dir in directions:
                    new_i = i +dir[0]
                    new_j = j + dir[1]
                    if ((0 <= new_i < len(grid)) and (0<= new_j < len(grid[0])) and grid[new_i][new_j] == 1):
                        grid[new_i][new_j] = 2
                        queue.append((new_i, new_j))
                        count -= 1
            minutes += 1   
        return max(0, minutes-1) if (count == 0) else -1